scripts/md-gen/controller-config/main.go

```diff
@@ -0,0 +1,472 @@
+		}
+		return rawType
+	}
+
+	callExpr := expr.(*ast.CallExpr)
+	rawType := callExpr.Fun.(*ast.SelectorExpr).Sel.Name
+	dataType := niceNameFor(rawType)
+
+	if len(callExpr.Args) > 0 {
+		// add parameter types
+		dataType += "["
+		for i, arg := range callExpr.Args {
+			if i > 0 {
+				dataType += ", "
+			}
+			dataType += typeForExpr(arg)
+		}
+		dataType += "]"
+	}
+
+	return dataType
+}
+
+// Check whether key is mutable in AllowedUpdateConfigAttributes slice
+func fillFromAllowedUpdateConfigAttributes(data map[string]*keyInfo) {
+	for key := range controller.AllowedUpdateConfigAttributes {
+		ensureDefined(data, key)
+		data[key].Mutable = true
+	}
+}
+
+// keys for which a default value doesn't make sense
+var skipDefault = set.NewStrings(
+	controller.AuditLogExcludeMethods, // "[ReadOnlyMethods]" - not useful
+	controller.CACertKey,
+	controller.ControllerUUIDKey,
+)
+
+// Get default values using reflection on controller.Config type
+func fillFromNewConfig(data map[string]*keyInfo) {
+	config, err := controller.NewConfig(testing.ControllerTag.Id(), testing.CACert, nil)
+	check(err)
+	for key, defaultVal := range config {
+		if skipDefault.Contains(key) {
+			continue
+		}
+
+		ensureDefined(data, key)
+		data[key].Default = fmt.Sprint(defaultVal)
+	}
+}
+
+// Get default values using reflection on controller.Config type
+// Used as a fallback where fillFromNewConfig can't produce a value
+func fillFromConfigType(data map[string]*keyInfo) {
+	// Don't get defaults from these methods - generally bogus values
+	skipMethods := set.NewStrings(
+		"CAASImageRepo",
+		"CAASOperatorImagePath",
+		"ControllerAPIPort",
+		"Features",
+		"IdentityPublicKey",
+		"Validate", // not a config key
+	)
+
+	constantNameForMethod := func(methodName string) string {
+		name := strings.TrimSuffix(methodName, "MB")
+
+		rename := map[string]string{
+			"NUMACtlPreference": "SetNUMAControlPolicyKey",
+		}
+		if rn, ok := rename[name]; ok {
+			name = rn
+		}
+
+		return name
+	}
+
+	config, err := controller.NewConfig(testing.ControllerTag.Id(), testing.CACert, nil)
+	check(err)
+	t := reflect.TypeOf(config)
+	v := reflect.ValueOf(config)
+
+	for i := 0; i < t.NumMethod(); i++ {
+		method := t.Method(i)
+		methodValue := v.Method(i)
+
+		if skipMethods.Contains(method.Name) {
+			continue
+		}
+		if method.Type.NumIn() == 1 {
+			defaultVal := methodValue.Call([]reflect.Value{})[0]
+
+			constantName := constantNameForMethod(method.Name)
+			key, ok := keyForConstantName[constantName]
+			if !ok {
+				// Try adding "Key" suffix
+				key, ok = keyForConstantName[constantName+"Key"]
+				if !ok {
+					panic(method.Name)
+				}
+			}
+			if skipDefault.Contains(key) {
+				continue
+			}
+
+			ensureDefined(data, key)
+			data[key].Default2 = fmt.Sprint(defaultVal)
+		}
+	}
+}
+
+// UTILITY FUNCTIONS
+
+// Return the first value that is defined / not-zero, and "true"
+// if such a value is found.
+func firstNonzero(vals ...string) (string, bool) {
```

Comment:
Interestingly, this is likely going to go into the stdlib soon as `cmp.Or` (generically typed). See https://github.com/golang/go/issues/60204